File: my_flutter_app/test/notification_routing_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:my_flutter_app/screens/review_composer_screen.dart';
import 'package:my_flutter_app/screens/tracking_screen.dart';
import 'package:my_flutter_app/services/notification_routing.dart';

void main() {
  businessRoutingTests();
  group('routeForNotificationData', () {
    test('routes a car purchase update to My purchases', () {
      final route = routeForNotificationData({
        'type': 'car_purchase_status',
        'purchaseId': 'abc123',
      });

      expect(route?.name, '/my-purchases');
    });

    test('routes a freight status update to Orders freight tab + record', () {
      final route = routeForNotificationData({
        'type': 'freight_shipment_status',
        'shipmentId': 'ship-1',
      });

      expect(route?.name, '/orders');
      final args = route?.arguments as OrdersScreenArguments;
      expect(args.innerTab, 'freight');
      expect(args.focusId, 'ship-1');
      expect(args.focusCollection, 'freightShipments');
    });

    test('routes a barrel status update to Orders barrels tab + record', () {
      final route = routeForNotificationData({
        'type': 'barrel_shipment_status',
        'shipmentId': 'barrel-9',
      });

      expect(route?.name, '/orders');
      final args = route?.arguments as OrdersScreenArguments;
      expect(args.innerTab, 'barrels');
      expect(args.focusId, 'barrel-9');
      expect(args.focusCollection, 'barrelShipments');
    });

    test('routes a freight balance-due and refund event to Orders', () {
      final balanceDue = routeForNotificationData({
        'type': 'freight_balance_due',
        'shipmentId': 'ship-2',
      });
      final refund = routeForNotificationData({
        'type': 'freight_refund_issued',
        'shipmentId': 'ship-2',
      });

      expect(balanceDue?.name, '/orders');
      expect(refund?.name, '/orders');
      expect(
        (balanceDue?.arguments as OrdersScreenArguments).innerTab,
        'freight',
      );
    });

    test('routes a parking reservation update to Orders cars tab', () {
      final route = routeForNotificationData({
        'type': 'parking_reservation_status',
        'relatedId': 'park-1',
      });

      expect(route?.name, '/orders');
      final args = route?.arguments as OrdersScreenArguments;
      expect(args.innerTab, 'cars');
      expect(args.focusId, 'park-1');
    });

    test('routes a transport request update to Orders transport tab', () {
      final route = routeForNotificationData({
        'type': 'transport_request_status',
        'requestId': 'tr-1',
      });

      expect(route?.name, '/orders');
      final args = route?.arguments as OrdersScreenArguments;
      expect(args.innerTab, 'transport');
      expect(args.focusId, 'tr-1');
    });

    test('routes a review request to the review composer', () {
      final route = routeForNotificationData({
        'type': 'review_request',
        'relatedCollection': 'parkedCars',
        'relatedId': 'park-1',
        'businessId': 'biz-1',
      });

      expect(route?.name, '/leave-review');
      final args = route?.arguments as ReviewComposerArguments;
      expect(args.relatedCollection, 'parkedCars');
      expect(args.relatedId, 'park-1');
      expect(args.businessId, 'biz-1');
    });

    test('does not route a review request missing required fields', () {
      final route = routeForNotificationData({
        'type': 'review_request',
        'relatedCollection': 'parkedCars',
      });

      expect(route, isNull);
    });

    test('does not route a retired wallet refund update', () {
      // The wallet is retired (docs/PLAN-2026-08-backlog.md #3): the backend
      // never sends this type and /wallet is not a registered route, so it
      // must resolve to null rather than a tap that goes nowhere.
      final route = routeForNotificationData({
        'type': 'wallet_refund_status',
      });

      expect(route, isNull);
    });

    test('routes a car viewing update to its own destination', () {
      // Deliberately not /my-purchases: a buyer opening a notification about
      // an appointment should not land in a list of money they have paid.
      final route = routeForNotificationData({'type': 'car_viewing_status'});

      expect(route?.name, '/my-viewings');
    });

    test('still routes a car purchase update to My Purchases', () {
      final route = routeForNotificationData({'type': 'car_purchase_status'});

      expect(route?.name, '/my-purchases');
    });

    test('routes a support case update to the support thread', () {
      final route = routeForNotificationData({
        'type': 'support_case_update',
        'caseId': 'case-123',
      });

      expect(route?.name, '/support-thread');
      expect(route?.arguments, 'case-123');
    });

    test('does not route a support case update with no case id', () {
      final route = routeForNotificationData({'type': 'support_case_update'});

      expect(route, isNull);
    });

    test('routes a shipment tracking update to Tracking', () {
      final route = routeForNotificationData({
        'type': 'shipment_tracking_update',
        'shipmentId': 'ship-9',
      });

      expect(route?.name, '/tracking');
      final args = route?.arguments as TrackingScreenArguments;
      expect(args.shipmentId, 'ship-9');
    });

    test('routes a business application update to Business registration',
        () {
      final route = routeForNotificationData({
        'type': 'business_application_status',
      });

      expect(route?.name, '/business-register');
    });

    test('routes a support message to the support thread with its case id',
        () {
      final route = routeForNotificationData({
        'type': 'support_message',
        'caseId': 'case-7',
      });

      expect(route?.name, '/support-thread');
      expect(route?.arguments, 'case-7');
    });

    test('routes a support escalation the same way as a support message', () {
      final route = routeForNotificationData({
        'type': 'support_escalated',
        'caseId': 'case-8',
      });

      expect(route?.name, '/support-thread');
      expect(route?.arguments, 'case-8');
    });

    test('does not route a support message with no case id', () {
      final route = routeForNotificationData({'type': 'support_message'});

      expect(route, isNull);
    });

    test('does not route an unrecognized or missing type', () {
      expect(routeForNotificationData({}), isNull);
      expect(routeForNotificationData({'type': 'something_new'}), isNull);
    });
  });

  group('notification payload encode/decode', () {
    test('round-trips a typical data payload', () {
      final data = {
        'type': 'freight_shipment_status',
        'shipmentId': 'ship-1',
        'status': 'in_transit',
      };

      final decoded = decodeNotificationPayload(
        encodeNotificationPayload(data),
      );

      expect(decoded, {
        'type': 'freight_shipment_status',
        'shipmentId': 'ship-1',
        'status': 'in_transit',
      });
    });

    test('round-trips values containing & and = characters safely', () {
      final data = {'type': 'x', 'note': 'a=b&c=d'};

      final decoded = decodeNotificationPayload(
        encodeNotificationPayload(data),
      );

      expect(decoded['note'], 'a=b&c=d');
    });

    test('decodes an empty payload to an empty map', () {
      expect(decodeNotificationPayload(''), <String, String>{});
    });
  });

  group('a price arriving', () {
    test('opens the request it belongs to', () {
      // This notification had no case at all: the tap resolved to null and
      // the customer was left on whatever screen they were on, with no way
      // back to the price they had just been told about.
      final route = routeForNotificationData({
        'type': 'freight_quote_received',
        'requestId': 'req-123',
        'trackingCode': 'FQ-Z6YPQT',
      });
      expect(route, isNotNull);
      expect(route!.name, '/freight-quote');
      final args = route.arguments as FreightQuoteScreenArguments;
      expect(args.requestId, 'req-123');
      expect(args.trackingCode, 'FQ-Z6YPQT');
    });

    test('goes nowhere without a request to open', () {
      expect(
        routeForNotificationData({'type': 'freight_quote_received'}),
        isNull,
      );
    });
  });
}

void businessRoutingTests() {
  group('routeForNotificationData for a business', () {
    const business = NotificationAudience.business;

    test('a paid parking link opens that parked car', () {
      final route = routeForNotificationData({
        'type': 'business_order_paid',
        'service': 'parking',
        'reservationId': 'car-1',
      }, audience: business);
      expect(route?.name, '/business-record');
      final args = route?.arguments as BusinessRecordArguments;
      expect(args.collection, 'parkedCars');
      expect(args.id, 'car-1');
      expect(args.fallbackCategory, 'parking');
    });

    test('a paid barrel or transport job opens its record', () {
      final barrel = routeForNotificationData({
        'type': 'business_order_paid',
        'service': 'barrel',
        'shipmentId': 'b-1',
      }, audience: business);
      expect((barrel?.arguments as BusinessRecordArguments).collection,
          'barrelShipments');
      final transport = routeForNotificationData({
        'type': 'business_order_paid',
        'service': 'transport',
        'requestId': 't-1',
      }, audience: business);
      expect((transport?.arguments as BusinessRecordArguments).collection,
          'transportRequests');
    });

    test('a transport opportunity opens the transport jobs screen', () {
      expect(
        routeForNotificationData({'type': 'transport_opportunity'},
            audience: business)?.name,
        '/business-transport',
      );
    });

    test('quote and pool notices land on the business home list', () {
      final freight = routeForNotificationData({'type': 'freight_quote_request'},
          audience: business);
      expect(freight?.name, '/business-home');
      expect((freight?.arguments as BusinessHomeArguments).category, 'freight');
      final pool = routeForNotificationData({'type': 'barrel_pool_join'},
          audience: business);
      expect((pool?.arguments as BusinessHomeArguments).category, 'barrels');
    });

    test('viewing requests, verification and support go to their screens', () {
      expect(routeForNotificationData({'type': 'car_viewing_status'},
              audience: business)?.name,
          '/purchase-management');
      expect(routeForNotificationData({'type': 'business_verification_review'},
              audience: business)?.name,
          '/business-profile');
      expect(routeForNotificationData({'type': 'business_support_request'},
              audience: business)?.name,
          '/business-support');
      expect(routeForNotificationData(
              {'type': 'support_message', 'caseId': 'c1'}, audience: business)
          ?.name,
          '/support-thread');
    });

    test('a customer still gets the customer screens for shared types', () {
      expect(routeForNotificationData({'type': 'car_viewing_status'})?.name,
          '/my-viewings');
      expect(
        routeForNotificationData({
          'type': 'parking_reservation_status',
          'reservationId': 'r1',
        })?.name,
        '/orders',
      );
    });
  });
}

File: admin_web/src/lib/notification-routing.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import test from "node:test";

import {
  adminTargetForNotification,
  businessTargetForNotification,
  collectionFromNotification,
  customerOrdersInnerTab,
  customerTargetForNotification,
  focusFromNotification,
} from "./notification-routing.ts";

test("a live barrel status payload with only shipmentId opens Barrels", () => {
  const data = {type: "barrel_shipment_status", shipmentId: "ship-1"};
  const target = customerTargetForNotification(data);
  assert.equal(target.tab, "orders");
  assert.deepEqual(target.focus, {collection: "barrelShipments", id: "ship-1"});
  assert.equal(customerOrdersInnerTab(target.focus!.collection), "barrels");
});

test("freight status, balance-due and refund open Freight, not Cars", () => {
  for (const type of [
    "freight_shipment_status",
    "freight_balance_due",
    "freight_refund_issued",
  ]) {
    const target = customerTargetForNotification({type, shipmentId: "fr-9"});
    assert.equal(target.tab, "orders", type);
    assert.deepEqual(target.focus, {
      collection: "freightShipments",
      id: "fr-9",
    });
    assert.equal(customerOrdersInnerTab("freightShipments"), "freight");
  }
});

test("a stamped relatedCollection wins over type inference", () => {
  const focus = focusFromNotification({
    type: "shipment_tracking_update",
    relatedCollection: "transportRequests",
    relatedId: "tr-1",
    shipmentId: "ignored",
  });
  assert.deepEqual(focus, {collection: "transportRequests", id: "tr-1"});
});

test("freight_quote_received opens Orders → Freight for that request", () => {
  const target = customerTargetForNotification({
    type: "freight_quote_received",
    requestId: "req-3",
  });
  assert.equal(target.tab, "orders");
  assert.deepEqual(target.focus, {
    collection: "freightQuoteRequests",
    id: "req-3",
  });
  assert.equal(customerOrdersInnerTab("freightQuoteRequests"), "freight");
});

test("transport_request_status opens Orders → Car transport", () => {
  const target = customerTargetForNotification({
    type: "transport_request_status",
    requestId: "tr-4",
  });
  assert.equal(target.tab, "orders");
  assert.deepEqual(target.focus, {
    collection: "transportRequests",
    id: "tr-4",
  });
  assert.equal(customerOrdersInnerTab("transportRequests"), "transport");
});

test("parking_reservation_status uses reservationId on Cars & parking", () => {
  const target = customerTargetForNotification({
    type: "parking_reservation_status",
    reservationId: "park-2",
  });
  assert.equal(target.tab, "orders");
  assert.deepEqual(target.focus, {collection: "parkedCars", id: "park-2"});
  assert.equal(customerOrdersInnerTab("parkedCars"), "cars");
});

test("car_purchase_status uses purchaseId on that purchase", () => {
  const target = customerTargetForNotification({
    type: "car_purchase_status",
    purchaseId: "buy-8",
  });
  assert.equal(target.tab, "orders");
  assert.deepEqual(target.focus, {collection: "carPurchases", id: "buy-8"});
});

test("car_viewing_status opens Viewings with the appointment id", () => {
  const target = customerTargetForNotification({
    type: "car_viewing_status",
    purchaseId: "view-1",
  });
  assert.equal(target.tab, "viewings");
  assert.deepEqual(target.focus, {collection: "carPurchases", id: "view-1"});
});

test("review_request opens the review composer for that record", () => {
  const target = customerTargetForNotification({
    type: "review_request",
    relatedCollection: "parkedCars",
    relatedId: "park-1",
    businessId: "biz-1",
  });
  assert.equal(target.tab, "orders");
  assert.equal(target.openReview, true);
  assert.deepEqual(target.focus, {collection: "parkedCars", id: "park-1"});
});

test("support_message and support_escalated open the case thread", () => {
  for (const type of ["support_message", "support_escalated"]) {
    const target = customerTargetForNotification({type, caseId: "case-7"});
    assert.equal(target.tab, "support", type);
    assert.equal(target.caseId, "case-7", type);
  }
});

test("support_case_update opens the support thread, not Home", () => {
  const target = customerTargetForNotification({
    type: "support_case_update",
    caseId: "case-9",
  });
  assert.equal(target.tab, "support");
  assert.equal(target.caseId, "case-9");
});

test("payment_hold_capture_notice opens the held order", () => {
  const target = customerTargetForNotification({
    type: "payment_hold_capture_notice",
    recordId: "hold-rec",
    orderType: "transport_job",
  });
  assert.equal(target.tab, "orders");
  assert.deepEqual(target.focus, {
    collection: "transportRequests",
    id: "hold-rec",
  });
});

test("secured_order_cancelled_by_business opens that booking", () => {
  const target = customerTargetForNotification({
    type: "secured_order_cancelled_by_business",
    orderType: "freightShipment",
    recordId: "fr-cancel",
  });
  assert.equal(target.tab, "orders");
  assert.deepEqual(target.focus, {
    collection: "freightShipments",
    id: "fr-cancel",
  });
});

test("an unknown customer type stays on Home", () => {
  assert.equal(customerTargetForNotification({type: "something_new"}).tab, "home");
  assert.equal(customerTargetForNotification({}).tab, "home");
});

test("business freight quote types open Freight price-requests", () => {
  for (const type of [
    "freight_quote_request",
    "freight_quote_won",
    "freight_quote_lost",
  ]) {
    const target = businessTargetForNotification({type, requestId: "fq-1"});
    assert.equal(target.tab, "freight", type);
    assert.equal(target.panelView, "requests", type);
    assert.equal(target.focusId, "fq-1", type);
  }
});

test("transport_job_paid and paid transport open Accepted jobs", () => {
  const paid = businessTargetForNotification({
    type: "transport_job_paid",
    requestId: "tr-paid",
  });
  assert.equal(paid.tab, "transport");
  assert.equal(paid.panelView, "jobs");
  assert.equal(paid.focusId, "tr-paid");

  const order = businessTargetForNotification({
    type: "business_order_paid",
    service: "transport",
    requestId: "tr-paid-2",
  });
  assert.equal(order.tab, "transport");
  assert.equal(order.panelView, "jobs");
  assert.equal(order.focusId, "tr-paid-2");
});

test("transport_quote_won opens jobs and lost opens opportunities", () => {
  const won = businessTargetForNotification({
    type: "transport_quote_won",
    requestId: "tr-w",
  });
  const lost = businessTargetForNotification({
    type: "transport_quote_lost",
    requestId: "tr-l",
  });
  assert.equal(won.panelView, "jobs");
  assert.equal(lost.panelView, "opportunities");
  assert.equal(won.tab, "transport");
  assert.equal(lost.tab, "transport");
});

test("business car_viewing_status opens Viewings, not Purchases", () => {
  const target = businessTargetForNotification({
    type: "car_viewing_status",
    purchaseId: "view-2",
  });
  assert.equal(target.tab, "viewings");
  assert.equal(target.focusId, "view-2");
});

test("business_order_paid focuses the record on the service tab", () => {
  const cases: Array<[string, string, string]> = [
    ["barrels", "shipmentId", "b-1"],
    ["freight", "shipmentId", "f-1"],
    ["parking", "reservationId", "p-1"],
    ["purchases", "purchaseId", "c-1"],
  ];
  for (const [service, field, id] of cases) {
    const target = businessTargetForNotification({
      type: "business_order_paid",
      service,
      [field]: id,
    });
    assert.equal(target.tab, service, service);
    assert.equal(target.focusId, id, service);
  }
});

test("business support types open the case thread", () => {
  for (const type of [
    "support_message",
    "support_escalated",
    "support_case_update",
  ]) {
    const target = businessTargetForNotification({type, caseId: "case-b"});
    assert.equal(target.tab, "cases", type);
    assert.equal(target.caseId, "case-b", type);
  }
});

test("admin support types open the Support thread", () => {
  for (const type of [
    "support_message",
    "support_escalated",
    "support_case_update",
  ]) {
    const target = adminTargetForNotification({type, caseId: "case-a"});
    assert.equal(target.tab, "support", type);
    assert.equal(target.caseId, "case-a", type);
  }
});

test("admin support_case_update is not Today", () => {
  assert.notEqual(
    adminTargetForNotification({type: "support_case_update", caseId: "x"}).tab,
    "today",
  );
});

test("collection inference covers live payloads without relatedCollection", () => {
  assert.equal(
    collectionFromNotification({type: "barrel_shipment_status", shipmentId: "1"}),
    "barrelShipments",
  );
  assert.equal(
    collectionFromNotification({
      type: "parking_reservation_status",
      reservationId: "2",
    }),
    "parkedCars",
  );
  assert.equal(
    collectionFromNotification({
      type: "payment_hold_capture_notice",
      orderType: "parking_deposit",
      recordId: "3",
    }),
    "parkedCars",
  );
});

test("the consoles call the shared router instead of a local switch", () => {
  const customer = readFileSync("src/components/customer-console.tsx", "utf8");
  const business = readFileSync("src/components/business-console.tsx", "utf8");
  const admin = readFileSync("src/components/admin-console.tsx", "utf8");
  const panels = readFileSync(
    "src/components/business/operations-panels.tsx",
    "utf8",
  );
  const support = readFileSync("src/components/customer-support.tsx", "utf8");
  const cases = readFileSync(
    "src/components/support/support-cases-panel.tsx",
    "utf8",
  );
  assert.match(customer, /customerTargetForNotification/);
  assert.match(customer, /caseId=\{focusCaseId\}/);
  assert.match(customer, /customerOrdersInnerTab\(focusedRecord\.collection\)/);
  assert.match(support, /caseId = ""/);
  assert.match(business, /businessTargetForNotification/);
  assert.match(business, /focusView=\{notificationFocusView\}/);
  assert.match(
    panels,
    /setView\(focusView === "requests" \? "requests" : "shipments"\)/,
  );
  assert.match(
    panels,
    /setView\(focusView === "jobs" \? "jobs" : "opportunities"\)/,
  );
  assert.match(admin, /adminTargetForNotification/);
  assert.match(admin, /caseId=\{notificationCaseId\}/);
  assert.match(cases, /caseId = ""/);
});
